Self-hosted site user management

crazytonyli opened the pull request:
The majority of code in this PR is taken from #23572. See the first comment.

I added some minor changes in the following two commits: Removing a couple of global static variables and adding a feature flag for the new User Management feature.

I will make more changes to fine-tune both the UI and the implementation.

## Regression Notes
1. Potential unintended areas of impact


2. What I did to test those areas of impact (or what existing automated tests I relied on)


3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding unit tests for my changes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

crazytonyli committed:
Add site users list to self-hosted sites

This code is taken from PR #23572, with minimal changes.

crazytonyli committed:
Remove static variable `UserObjectResolver`

crazytonyli committed:
Add a feature flag for self-hosted site user management

dangermattic commented:
<!--
  0 Errors
  3 Warnings: <code>Modules/Package.swift</c...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages as appropriate to your project setup (e.g. in Xcode or by running <code>swift package resolve</code>).</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-iOS/milestone/303">25.6</a>. This milestone is due in less than 4 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

crazytonyli committed:
Add a release note

kean commented:
- Email is displayed in two lines – looks off
- Profile picture low quality
- I just deleted myself and it was the only user on my local tests site, so can no longer login in my site 🤣 I think it needs a bit more confirmation/alerts if you are deleting yourself.

crazytonyli commented:
Yep. I have a TODO list for those issues. As I mentioned in the PR description, they'll be addressed in future PRs.

> I just deleted myself 

🤣 I applaud your adventurous spirit! I saw the button but didn't have the guts to tap it. I'm surprised REST API would allow you to do that...

kean commented on Modules/Sources/WordPressShared/Utility/StringRankedSearch.swift:
```diff
@@ -120,3 +120,14 @@ extension StringRankedSearch {
 }
+
+/// Objects conforming to `StringRankedSearchable` can be searched by calling `search(query:)` on a collection of them
+public protocol StringRankedSearchable {
```
I'd suggest removing the protocol and sticking with closures. The same type can have different search criteria depending on the context, so passing a closure is more flexible. This is similar to the way `filter(:)` uses closures and not protocols. 

It may be worth adding a generic to `Collection` or something like this, but I'm not sure it's worth it.

```
extension Collection {
    func search(searchTerm: query, _ closure: (Element) -> String)
}
```

kean commented on Modules/Sources/WordPressShared/Utility/StringRankedSearch.swift:
```diff
@@ -120,3 +120,14 @@ extension StringRankedSearch {
+    var searchString: String { get }
+}
+
+public extension Collection where Iterator.Element: StringRankedSearchable {
```
(nit) I think `Collection` should have `Element`, so no need to reach for the `Iterator`

kean commented on Modules/Sources/WordPressUI/Views/Users/Components/UserListItem.swift:
```diff
@@ -0,0 +1,42 @@
+    @ScaledMetric(relativeTo: .headline)
+    var height: CGFloat = 48
+
+    @Environment(\.sizeCategory)
```
(nit) it is deprecated https://developer.apple.com/documentation/swiftui/environmentvalues/sizecategory and the [replacement](https://developer.apple.com/documentation/swiftui/environmentvalues/dynamictypesize) is iOS 15+

kean commented on Modules/Sources/WordPressUI/Views/Users/Components/UserListItem.swift:
```diff
@@ -0,0 +1,42 @@
+    private let userProvider: UserDataProvider
+    private let actionDispatcher: UserManagementActionDispatcher
+
+    init(user: DisplayUser, userProvider: UserDataProvider, actionDispatcher: UserManagementActionDispatcher) {
```
(nit) I usually take advantage of the auto-generated init for structs by making the required parameters `internal`.

kean commented on Modules/Sources/WordPressUI/Views/Users/UserProvider.swift:
```diff
@@ -0,0 +1,45 @@
+/// to perform user management actions.
+///
+/// The default implementation is set up for testing with SwiftUI Previews
+open class UserManagementActionDispatcher: ObservableObject {
```
The app doesn't use this type of naming elsewhere ("Dispatcher"). I'd suggest sticking with the current naming conventions. I would also suggest avoiding subclassing by adding a protocol like this:

```swift
protocol UserManagementServiceProtocol {
    // Changed from "fetchCurrentUserCan" and made non-async (we aren't going to perform a fetch for every individual capability?)
    func hasCapability(_ capability: String) -> Bool

    // Or something else other that a separate CachedUserListCallback closure
    func getCachedUsers() -> [User]

    func fetchUsers() async throws -> [WordPressUI.DisplayUser]
    
    // Why was it async throws? Noting that it seems to be unused.
    func invalidateCaches() 

    func setNewPassword(id: Int32, newPassword: String) async throws

    func deleteUser(id: Int32, reassigningPostsTo userId: Int32) async throws
}
```

kean replied:
It's also not clear what `WordPressUI.DisplayUser` is and what `UserDataProvider` belongs to.

If it's a model layer, it should returns users as is: `User`, `SiteUser`?. Perhaps there needs to be a namespace for the types associated with the WordPress sites to disambiguate. The name `DisplayUser` doesn't make much sense. 

If it's a ViewModel layer, it should be `UserView(Cell)Model`.

kean commented on Modules/Sources/WordPressUI/Views/Users/ViewModel/DisplayUser.swift:
```diff
@@ -0,0 +1,76 @@
+
+        // These are in ranked order – the higher something is in the list, the more heavily it's weighted
+        [
+            handle,
```
The main things that should be on this list are `username` and `displayName`, which is what you see on the "Users" list. It _may_ be worth also adding a `emailAddress`. 

`emailAddress` appears twice.

kean commented on Modules/Sources/WordPressUI/Views/Users/Components/UserProfileImage.swift:
```diff
@@ -0,0 +1,37 @@
+import SwiftUI
+
+struct UserProfileImage: View {
```
Please, use `AvatarView` and avoid `AsyncImage` as it uses `URLSession.shared` and basically has no caching. `AvatarView` is in the app target, so `User*` types will likely need to be moved to the app target.

The app has `CachedAsyncImage` that uses app's `ImageDownloader` with a memory cache.

kean commented on Modules/Sources/WordPressUI/Views/Users/Components/UserProfileImage.swift:
```diff
@@ -0,0 +1,37 @@
+
+    init(size: CGSize, email: String) {
+        self.size = size
+        self.url = URL(string: "https://gravatar.com/avatar/58fc51586c9a1f9895ac70e3ca60886e?size=256")!
```
(nit) mock?

kean commented on Modules/Sources/WordPressUI/Views/Users/ViewModel/UserChangePasswordViewModel.swift:
```diff
@@ -0,0 +1,76 @@
+    var password: String = ""
+
+    @Published
+    var isChangingPassword: Bool = false
```
(nit) Add `private(set)` where it applies.

kean commented on Modules/Sources/WordPressUI/Views/Users/ViewModel/UserChangePasswordViewModel.swift:
```diff
@@ -0,0 +1,76 @@
+        }
+
+        Task {
+            await MainActor.run {
```
You typically don't want to use `await MainActor.run` imperatively. The better option is to declare that `Task` runs is isolated to the main actor:

```swift
Task { @MainActor in
   // remove await MainActor.run
}
```

I'd also consider isolating the entire ViewModel to the main actor to ensure none of its properties are updated in the background.

kean commented on Modules/Sources/WordPressUI/Views/Users/ViewModel/UserChangePasswordViewModel.swift:
```diff
@@ -0,0 +1,76 @@
+            }
+
+            await MainActor.run {
+                callback()
```
The `callback` isn't needed since the ViewModel already has `dismissAction` and can dismiss itself.

kean commented on Modules/Sources/WordPressUI/Views/Users/ViewModel/UserDeleteViewModel.swift:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    func fetchOtherUsers() async {
+        withAnimation {
```
(nit) It works, but I'd suggest looking into [animation(_:value:)](https://developer.apple.com/documentation/swiftui/view/animation(_:value:)) for specifying animations in the view layer.

kean replied:
+ same comments are about `private(set)` and `@MainActor` in task apply here.

kean commented on WordPress/Classes/ViewRelated/Blog/Blog Details/BlogDetailsViewController.m:
```diff
@@ -833,6 +839,20 @@ - (BlogDetailsRow *)peopleRow
+                                        accessibilityIdentifier:@"User Row"
+                                                          image:[UIImage imageNamed:@"site-menu-people"]
+                                                       callback:^{
+        if (@available(iOS 16.4, *)) {
```
It doesn't seem there is anything preventing it from running on earlier versions.

kean commented on WordPress/Classes/Services/UserService.swift:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    func profilePhotoUrl(for user: UserWithEditContext) -> URL? {
+        guard let rawUrl = user.avatarUrls?.first?.value else { // This results in a very low-res avatar
```
(nit) "This results in a very low-res avatar " – it sure does 😄

kean commented on WordPress/Classes/Services/UserService.swift:
```diff
@@ -0,0 +1,111 @@
+             }
+
+             return DisplayUser(
+                 id: $0.id,
```
(nit) I'd suggest moving it to a convenience initializer.

kean commented on Modules/Sources/WordPressUI/Views/Users/Views/UserListView.swift:
```diff
@@ -0,0 +1,72 @@
+    public var body: some View {
+        Group {
+            if let error = viewModel.error {
+                EmptyStateView(error.localizedDescription, systemImage: "exclamationmark.triangle.fill")
```
I'd suggest using existing factory method `EmptyStateView.failure` with a "retry" button.

kean commented on Modules/Sources/WordPressUI/Views/Users/Views/UserDetailView.swift:
```diff
@@ -0,0 +1,178 @@
+
+    enum Strings {
+        static let nameSectionTitle = NSLocalizedString(
+            "userdetail.name-section-title",
```
These need to use cameCase https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/docs/localization.md#use-unique-reverse-dns-naming-style-keys: `userDetails.nameSectionTitle".

(nit) I'd also suggesting keeping the localizable strings as one liners. You don't really need to make them readable and they take too much vertical space otherwise.

kean commented on Modules/Sources/WordPressUI/Views/Users/Views/UserDetailView.swift:
```diff
@@ -0,0 +1,178 @@
+import SwiftUI
+
+struct UserDetailView: View {
```
(nit) It should probably say `UserDetails` not "detail".

kean commented on Modules/Sources/WordPressUI/Views/Users/Views/UserDetailView.swift:
```diff
@@ -0,0 +1,178 @@
+    var body: some View {
+        Form {
+            Section(Strings.nameSectionTitle) {
+                LabeledContent(Strings.roleFieldTitle, value: user.role)
```
TIL didn't know about `LabeledContent`, nice.

kean commented on Modules/Sources/WordPressUI/Views/Users/Views/UserDeleteView.swift:
```diff
@@ -0,0 +1,70 @@
+                } label: {
+                    HStack {
+                        Spacer()
+                        Text("Delete User")
```
Missing localization (in the whole file).

kean commented on Modules/Sources/WordPressUI/Views/Users/Views/UserDetailView.swift:
```diff
@@ -0,0 +1,178 @@
+                    }
+
+                    NavigationLink {
+                        // Pass this view's dismiss action, because if we delete a user, we want that screen *and* this one gone
```
Both of these actions – "Set New Password" and "Delete User" should be shown as modal screens.
"Delete User" should be destructive.

kean commented on Modules/Sources/WordPressUI/Views/Users/Views/UserDetailView.swift:
```diff
@@ -0,0 +1,178 @@
+            }
+
+            Section(Strings.contactInfoSectionTitle) {
+                LabeledContent(Strings.emailAddressFieldTitle, value: user.emailAddress)
```
Probably not the best way to display it. I'd suggest a vertical stack.
<img width="465" alt="Screenshot 2024-11-06 at 11 49 50 AM" src="https://github.com/user-attachments/assets/b93fd4fe-3b6b-4297-a839-bbdfb9d2c2ef">

kean commented on Modules/Sources/WordPressUI/Views/Users/Views/UserChangePasswordView.swift:
```diff
@@ -0,0 +1,62 @@
+                }, label: {
+                    HStack {
+                        Spacer()
+                        Text("Save Changes")
```
Localization (entire file).

crazytonyli commented:
@kean Thanks for the review! I have addressed most of your comments [in their own commits](https://github.com/wordpress-mobile/WordPress-iOS/pull/23777#commits-pushed-7f592a3) at https://github.com/wordpress-mobile/WordPress-iOS/pull/23777. A couple of things that I didn't address because I plan to refactor them later
* UserService, UserDataProvider, and ActionDispatcher.
* The set new password and the delete user flows.
